components/provisioning/org.wso2.carbon.identity.provisioning/src/main/java/org/wso2/carbon/identity/provisioning/listener/ProvisioningRoleMgtListener.java

```diff
@@ -0,0 +1,124 @@
+import org.wso2.carbon.identity.role.v2.mgt.core.model.RoleBasicInfo;
+import org.wso2.carbon.user.api.UserStoreException;
+import org.wso2.carbon.user.core.UserCoreConstants;
+import org.wso2.carbon.user.core.common.AbstractUserStoreManager;
+
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * The implementation for outbound user provisioning based on role.
+ */
+public class ProvisioningRoleMgtListener extends AbstractRoleManagementListener {
+
+    private static final Log LOG = LogFactory.getLog(ProvisioningRoleMgtListener.class);
+
+    @Override
+    public boolean isEnable() {
+
+        IdentityEventListenerConfig identityEventListenerConfig = IdentityUtil.readEventListenerProperty
+                (RoleManagementListener.class.getName(), this.getClass().getName());
+        if (identityEventListenerConfig == null) {
+            return true;
+        }
+        return Boolean.parseBoolean(identityEventListenerConfig.getEnable());
+    }
+
+    @Override
+    public int getExecutionOrderId() {
+
+        IdentityEventListenerConfig identityEventListenerConfig = IdentityUtil.readEventListenerProperty
+                (RoleManagementListener.class.getName(), this.getClass().getName());
+        if (identityEventListenerConfig == null) {
+            return IdentityCoreConstants.EVENT_LISTENER_ORDER_ID;
+        }
+        return identityEventListenerConfig.getOrder();
+    }
+
+    @Override
+    public int getDefaultOrderId() {
+
+        IdentityEventListenerConfig identityEventListenerConfig = IdentityUtil.readEventListenerProperty
+                (RoleManagementListener.class.getName(), this.getClass().getName());
+        if (identityEventListenerConfig == null) {
+            return IdentityCoreConstants.EVENT_LISTENER_ORDER_ID;
+        }
+        return identityEventListenerConfig.getOrder();
+    }
+
+    @Override
+    public void postUpdateUserListOfRole(String roleId, List<String> newUserIDList, List<String> deletedUserIDList,
+                                         String tenantDomain) throws IdentityRoleManagementException {
+
+        try {
+            AbstractUserStoreManager userStoreManager = (AbstractUserStoreManager) ProvisioningServiceDataHolder
+                    .getInstance().getRealmService().getTenantUserRealm(IdentityTenantUtil.getTenantId(tenantDomain))
+                    .getUserStoreManager();
+
+            String[] deletedUserNames = resolveUserNameFromUserIds(deletedUserIDList, userStoreManager);
+
+            String[] newUserNames = resolveUserNameFromUserIds(newUserIDList, userStoreManager);
+
+            if (deletedUserNames.length > 0 && newUserNames.length > 0) {
+                return;
+            }
+
+            RoleBasicInfo roleBasicInfo = ProvisioningServiceDataHolder.getInstance().getRoleManagementService()
+                    .getRoleBasicInfoById(roleId,tenantDomain);
+            // Only organization audience roles are supported for role based outbound provisioning
```

Comment:
Shouldn't this be backwards compatible?

Reply:
Before IS 7.0 all the roles are organization audience roles. Hence previous roles and new roles created with organization audience role can be used for role base outbound provisioning.

The new type of role which is application audience roles came with IS 7.0. How to handle the role based provisioning for those types of roles should be analyzed due to below concerns.
- Role name are not unique hence we have to register role uuid or role prefixed with its corresponding app name
- As outbound provisioning is organization wide config by default (we can configure app wise also), the use cases of having application roles for role based outbound provisioning required to be analyzed.

Due to above concerns, outbound provisioning for application audience is not given at first phase. 

Hope the explanation is clear enough, please raise if there any concerns. Here backward compatibility is preserved as here the application roles means not the `Application/<role-name>` role type had in previous IS versions.